lib/events/events.go

```diff
@@ -319,12 +315,19 @@ func (s *bufferedSubscription) pollingLoop() {
 	}
 }
 
-func (s *bufferedSubscription) Since(id int, into []Event) []Event {
+func (s *bufferedSubscription) Since(id int, into []Event, timeout time.Duration) []Event {
 	s.mut.Lock()
 	defer s.mut.Unlock()
 
-	for id >= s.cur {
-		s.cond.Wait()
+	// Check once first before generating the TimeoutCondWaiter
+	if id >= s.cur {
+		waiter := s.cond.SetupWait(timeout)
+		for id >= s.cur {
```

Comment:
From the commit message:

> This gets a bit messy, as the timeout's Timer needs to span calls to Wait()

If we're waiting for an event in the future, we need to keep waiting until we see the event we're after. There's one timeout which spans all of these waits.